_posts/2019-03-11-dask-nep18.md

```diff
@@ -0,0 +1,521 @@
+we import libraries and create arrays:
+
+```python
+import numpy as np
+from dask_glm.estimators import LinearRegression
+import matplotlib.pyplot as plt
+
+N = 1000
+
+# x from 0 to N
+x = N * np.random.random((40000, 1))
+
+# y = a*x + b with noise
+y = 0.5 * x + 1.0 + np.random.normal(size=x.shape)
+
+# create a linear regression model
+est = LinearRegression(solver='lbfgs')
+```
+
+The rest of the code and also the plot look alike the previous scikit-learn
+example, so we're ommitting those here for brevity. Note also that we could
+have called `LinearRegression()` without any arguments, but for this example
+we chose the
+[`lbfgs`](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html)
+solver, that converges reasonably fast.
+
+We can also have a look at the timing results for fitting, followed by those
+for predicting with Dask-GLM:
+
+```
+# Fitting
+CPU times: user 9.66 ms, sys: 116 µs, total: 9.78 ms
+Wall time: 8.94 ms
+
+# Predicting
+CPU times: user 130 µs, sys: 327 µs, total: 457 µs
+Wall time: 1.06 ms
+```
+
+If instead we want to use CuPy to compute, we have to change only 3 lines,
+importing `cupy` instead of `numpy`, and the two lines where we create the
+random arrays, replacing them to `cupy.random` insted of `np.random`. The
+block should then look like this:
+
+```python
+import cupy
+from dask_glm.estimators import LinearRegression
+import matplotlib.pyplot as plt
+
+N = 1000
+
+# x from 0 to N
+x = N * cupy.random.random((40000, 1))
+
+# y = a*x + b with noise
+y = 0.5 * x + 1.0 + cupy.random.normal(size=x.shape)
+
+# create a linear regression model
+est = LinearRegression(solver='lbfgs')
+```
+
+And the timing results we obtain in this scenario are:
+
+```
+# Fitting
+CPU times: user 151 ms, sys: 40.7 ms, total: 191 ms
+Wall time: 190 ms
+
+# Predicting
+CPU times: user 1.91 ms, sys: 778 µs, total: 2.69 ms
+Wall time: 1.37 ms
+```
+
+For the simple example chosen for this post, scikit-learn outperforms Dask-GLM
+using both NumPy and CuPy arrays. There may exist several reasons that
+contribute to this, and while we didn't dive deep into understanding the exact
+reasons and their extent, we could cite some likely possibilities:
+
+*   scikit-learn may be using solvers that converge faster;
+*   Dask-GLM is entirely built on top of Dask, while scikit-learn may be
+    heavily optimized internally;
+*   Too many synchronization steps and data transfer could occur for small
+    datasets with CuPy.
+
+### Performance for different Dask-GLM solvers
+
+To verify how Dask-GLM with Dask arrays would compare with CuPy arrays, we also
+did some linear regression benchmarking of Dask-GLM solvers. The results below
+were obtained from a training dataset with 1000 features of 100 dimensions,
+and 1000 test features.
+
+```
+Solver admm with Dask took 22444.307 ms to fit and 41.299 ms to predict
+Solver admm with CuPy took 263.742 ms to fit and 1.146 ms to predict
```

Comment:
You might consider including a `max_iter=` keyword when solving.  According to @cicdw our stopping criteria in dask-glm need some work.

Reply:
Yes, you're right, but in this case we're comparing only the backends, so it will probably not make a big difference, thought it would could make a bigger difference when comparing to scikit-learn. But for a scikit-learn comparison, it could seem like cheating, since we can't set the same parameter there.

Reply:
Yea, stopping criteria for `proximal_grad` and `admm` are always finnicky; you might also want to include some sort of comparison between the actual outputs of the various model fits to ensure that they've all converged to something meaningful.

Reply:
Huh, the classes in `dask_glm.linear_models` don't have an `n_iter_` parameter. https://github.com/dask/dask-glm/issues/77

Reply:
I'm marking this as resolved, since this arises from the (possible) bug we identified, plus the new graphs I added give us some nice visualization of the current state between NumPy and CuPy.